plugin/src/main/scala-2.12/chisel3/internal/plugin/ChiselPlugin.scala

```diff
@@ -131,30 +175,47 @@ class ChiselComponent(val global: Global) extends PluginComponent with TypingTra
     override def transform(tree: Tree): Tree = tree match {
       // Check if a subtree is a candidate
       case dd @ ValDef(mods, name, tpt, rhs) if okVal(dd) =>
+        val tpe = inferType(tpt)
       // If a Data and in a Bundle, just get the name but not a prefix
-        if (shouldMatchData(dd) && inBundle(dd)) {
+        if (shouldMatchData(tpe) && inBundle(dd)) {
           val str = stringFromTermName(name)
           val newRHS = transform(rhs)  // chisel3.internal.plugin.autoNameRecursively
-          val named = q"chisel3.internal.plugin.autoNameRecursively($str, $newRHS)"
+          val named = q"chisel3.internal.plugin.autoNameRecursively($str)($newRHS)"
           treeCopy.ValDef(dd, mods, name, tpt, localTyper typed named)
         }
         // If a Data or a Memory, get the name and a prefix
-        else if (shouldMatchDataOrMem(dd)) {
+        else if (shouldMatchDataOrMem(tpe)) {
           val str = stringFromTermName(name)
           val newRHS = transform(rhs)
           val prefixed = q"chisel3.experimental.prefix.apply[$tpt](name=$str)(f=$newRHS)"
-          val named = q"chisel3.internal.plugin.autoNameRecursively($str, $prefixed)"
+          val named = q"chisel3.internal.plugin.autoNameRecursively($str)($prefixed)"
           treeCopy.ValDef(dd, mods, name, tpt, localTyper typed named)
         // If an instance, just get a name but no prefix
-        } else if (shouldMatchModule(dd)) {
+        } else if (shouldMatchModule(tpe)) {
           val str = stringFromTermName(name)
           val newRHS = transform(rhs)
-          val named = q"chisel3.internal.plugin.autoNameRecursively($str, $newRHS)"
+          val named = q"chisel3.internal.plugin.autoNameRecursively($str)($newRHS)"
           treeCopy.ValDef(dd, mods, name, tpt, localTyper typed named)
         } else {
           // Otherwise, continue
           super.transform(tree)
         }
+      case dd @ ValDef(mods, name, tpt, rhs @ Match(_, _)) if okUnapply(dd) =>
+        val tpe = inferType(tpt)
+        val fieldsOfInterest: List[Boolean] = tpe.typeArgs.map(shouldMatchData)
+        // Only transform if at least one field is of interest
+        if (fieldsOfInterest.reduce(_ || _)) {
```

Comment:
Is this never empty?

Reply:
If none of the fields are `<: Data` then it will be empty, eg.
```scala
val (a, b) = ("hi", 3)
```